asserts/user.go

```diff
@@ -0,0 +1,212 @@
+	if valid && !su.until.IsZero() {
+		valid = when.Before(su.until)
+	}
+	return valid
+}
+
+// Implement further consistency checks.
+func (su *SystemUser) checkConsistency(db RODatabase, acck *AccountKey) error {
+	// Do the cross-checks when this assertion is actually used,
+	// i.e. in the create-user code. See also Model.checkConsitency
+
+	return nil
+}
+
+// sanity
+var _ consistencyChecker = (*SystemUser)(nil)
+
+func checkHashedPassword(headers map[string]interface{}, name string) (string, error) {
+	pw, err := checkOptionalString(headers, name)
+	if err != nil {
+		return "", err
+	}
+	// crypt(3) compatible hashes have the form: $id$salt$hash
+	l := strings.SplitN(pw, "$", 5)
+	if len(l) != 4 && len(l) != 5 {
+		return "", fmt.Errorf(`%q header must be a hashed password of the form "$integer-id$salt$hash", see crypt(3)`, name)
+	}
+	// see crypt(3), ID 6 means SHA-512 (since glibc 2.7)
+	ID, err := strconv.Atoi(l[1])
+	if err != nil {
+		return "", fmt.Errorf(`%q header must start with "$integer-id$", got %q`, name, l[1])
+	}
+	// double check that we only allow modern hashes
+	if ID < 6 {
+		return "", fmt.Errorf("%q header only supports $id$ values of 6 (sha512crypt) or higher", name)
+	}
+
+	// the $rounds=N$ part is optional
+	i := 2
+	if strings.HasPrefix(l[i], "rounds=") {
+		i++
+	}
+
+	// see crypt(3) for the legal chars
+	validSaltAndHash := regexp.MustCompile(`^[a-zA-Z0-9./=]+$`)
+	if !validSaltAndHash.MatchString(l[i]) {
+		return "", fmt.Errorf("%q header has invalid chars in salt %q", name, l[2])
+	}
+	i++
+	if i >= len(l) {
+		return "", fmt.Errorf("%q header has missing hash field", name)
+	}
+	if !validSaltAndHash.MatchString(l[i]) {
+		return "", fmt.Errorf("%q header has invalid chars in hash %q", name, l[3])
+	}
+
+	return pw, nil
+}
+
+func assembleSystemUser(assert assertionBase) (Assertion, error) {
+	err := checkAuthorityMatchesBrand(&assert)
+	if err != nil {
+		return nil, err
+	}
+	email, err := checkNotEmptyString(assert.headers, "email")
+	if err != nil {
+		return nil, err
+	}
+	if _, err := mail.ParseAddress(email); err != nil {
+		return nil, fmt.Errorf(`"email" header must be a RFC 5322 compliant email address: %s`, err)
+	}
+
+	series, err := checkStringList(assert.headers, "series")
+	if err != nil {
+		return nil, err
+	}
+	models, err := checkStringList(assert.headers, "models")
+	if err != nil {
+		return nil, err
+	}
+	if _, err := checkOptionalString(assert.headers, "name"); err != nil {
+		return nil, err
+	}
+	if _, err := checkStringMatches(assert.headers, "username", validSystemUserUsernames); err != nil {
+		return nil, err
+	}
+	if _, err := checkHashedPassword(assert.headers, "password"); err != nil {
+		return nil, err
+	}
+
+	sshKeys, err := checkStringList(assert.headers, "ssh-keys")
+	if err != nil {
+		return nil, err
+	}
+	since, err := checkRFC3339Date(assert.headers, "since")
+	if err != nil {
+		return nil, err
+	}
+	until, err := checkRFC3339Date(assert.headers, "until")
```

Comment:
To start on the conservative side, can we please enforce "until" to be both present and not more than 367 days after "since"? We don't want brands to be signing off on eternal rooting for their devices.